### DISCUSSION TITLE
Interaction with the app through one device gets reflected on all other devices using the app

### DISCUSSION THREAD

**Discussion by mnobeidat13** [flet-dev/flet#665] [Q&A]
I deployed my app to replit.com. I open the app through two devices, my phone and my laptop. What happens is any interaction I do on either device gets reflected on the app opened in the other one. 
I cant explain more and screenshots wont help any further I guess. 
I dont know, but it seems that the app runs in one session.
I dont use any authintication nor I save user session (I font know if that is even possible).

This is a link to the app deployed on replit:
https://portfolio.mnobeidat.repl.co/#/item_recommendation

**Comment by FeodorFitsner** [comment]
Forgot link?

**Reply by mnobeidat13** [reply]
https://portfolio.mnobeidat.repl.co/#/item_recommendation

**Comment by FeodorFitsner** [comment]
I don't think I can see your interactions. Can you see mine session?

Don you call `flet.app()` or `flet.page()` to start the app?

**Reply by mnobeidat13** [reply]
When you start the app
In the project "recommender system" 
The interface should be clear at first
If its showing any results, it means that it has it from another session.

I am using flet.app()

**Reply by FeodorFitsner** [reply]
That could mean there is a shared (static) data outside a session scope (`main()`) method. Could tell more after looking into sources if the project is not private.

**Reply by mnobeidat13** [reply]
![Uploading Screenshot_20221204-043949_Chrome.jpg…]()

**Reply by mnobeidat13** [reply]
I can share the code, but there is more than 10 files and the code is not documented.
Can you explain more what you mean by static data?
Note: When I run the app in vscode multiple time, I get the same results, the output from the previous session, unless  I restart the notebook. 
That does not happen when I run the app through the command line.

**Reply by mnobeidat13** [reply]
Here is how I navigate through the app.
The view is built once the route changes.
If I clear the view controls and render them again on each route change, that might work.

![image](https://user-images.githubusercontent.com/32016587/205468406-e3fa5f4b-ce70-4611-bf1f-f920c2d9a8cc.png)

**Comment by FeodorFitsner** [comment]
See, here I'm using a shared `AppData` data between sessions:

```python
import flet as ft


class AppData:
    search_results = "one"


data = AppData()


def main(page: ft.Page):

    results = ft.Text()
    query = ft.TextField()

    def display_search():
        results.value = f"Search results: {data.search_results}"
        page.update()

    def perform_search(e):
        data.search_results = query.value
        query.value = ""
        display_search()

    page.add(
        ft.Row([query, ft.ElevatedButton("Search", on_click=perform_search)]), results
    )
    display_search()


ft.app(target=main, port=8550, view=ft.WEB_BROWSER)

```

Run this example and in any new session (new tab or page in the browser) you'll be seeing the last search result.

**Reply by FeodorFitsner** [reply]
Now, to fix that I'm creating an instance of `AppData` inside `main()` method:

```python
import flet as ft


class AppData:
    search_results = "one"


def main(page: ft.Page):

    data = AppData()

# ...
# all the rest is the same
```

**Reply by mnobeidat13** [reply]
I think I was doing something similar to this.
"The view is built once the route changes.
If I clear the view controls and render them again on each route change, that might work."
I tried this and it worked. I rebuild the view on route change.
But I still something cray happening.
I interact with the app through my laptop and the action happens on my phone and vice versa.

**Reply by FeodorFitsner** [reply]
It would be great to have a simplified, public repro, so I can investigate.

**Reply by mnobeidat13** [reply]
You can see in the video what is happening.
I have two instances of the app, on in replit webview and one in chrome 
When I interact with the app in chrome window, the interaction is reflected on the one inside replit

https://user-images.githubusercontent.com/32016587/205487491-62181e4b-ee9c-4067-b424-05059ceaa986.mp4

**Reply by FeodorFitsner** [reply]
It looks like some data is shared between sessions. Can't tell what's wrong without looking into sources unless you can do a repro.

**Reply by mnobeidat13** [reply]
Here I am running two instances of the app on my laptop and they dont interfere with each other.
I will document my code and publish it so you can take a look at it if that is possible.
I thick the problem is with replit, when you run an app it only creates on instance. But what I need is a host that creates a new instance of the app upon each request.
Is it possible to expose the app to a public IP address when running in Windows so I can do more testing?

https://user-images.githubusercontent.com/32016587/205515580-c323a609-99c4-4afd-99d7-d784ea9104ea.mp4

**Reply by FeodorFitsner** [reply]
Try deploying it to fly.io (they have a free tier) and see if the app behaves differently there: https://flet.dev/docs/guides/python/deploying-web-app/fly-io

I might remember seeing weird behavior on Replit before, but I didn't figure out how to describe/troubleshoot it...

**Reply by mnobeidat13** [reply]
Will do.
I really appreciate your interest and quick replies. Thanks a lot.

**Reply by mnobeidat13** [reply]
Anything about exposing the app to a public IP address?

**Reply by FeodorFitsner** [reply]
Sure, you can. By default `flet.app()` starts web server listening on all NIC IPs. Just open a port on your Windows firewall and forward on NAT/router :) Seriously, I don't recommend exposing your computer to the internet. Instead, you can use reverse proxy tool/service such as [ngrok](https://ngrok.com/).

**Comment by mnobeidat13** [comment]
I will see what I can do and if does not work i will share the code with you

**Comment by mnobeidat13** [comment]
@FeodorFitsner Hello again
I uploaded the files to this repository and tried to explain the structure of the app in the read me file 
https://github.com/mnobeidat13/portfolio
If you still need further explanation, let me know
the app is running on replit.com: https://replit.com/@mnobeidat/portfolio

I also tried running it on fly.io but the docker image was too big for the free tier
